### PR TITLE
Feat: Add Model RF-DETR

### LINKED ISSUE #327: [Object Detection Model] Please implement RF-DETR
RF-DETR is a Object Detection Model based on the [RF-DETR: Neural Architecture Search for Real-Time Detection Transformers](https://arxiv.org/abs/2511.09554) paper.

Please implement

Image

based [on the roboflow implementation](https://github.com/roboflow/rf-detr/)
with the pretrained weight available in 
 > https://torch-cdn.mlverse.org/models/vision/v2/models/rf_detr_nano.pth.
 > https://torch-cdn.mlverse.org/models/vision/v2/models/rf_detr_small.pth.
 > https://torch-cdn.mlverse.org/models/vision/v2/models/rf_detr_medium.pth.
 > https://torch-cdn.mlverse.org/models/vision/v2/models/rf_detr_base.pth.
 > https://torch-cdn.mlverse.org/models/vision/v2/models/rf_detr_base_2.pth.
 > https://torch-cdn.mlverse.org/models/vision/v2/models/rf_detr_base_o365.pth.
 > https://torch-cdn.mlverse.org/models/vision/v2/models/rf_detr_large.pth.

It is dependant of https://github.com/mlverse/torchvisionlib/issues/25

### PR DESCRIPTION
**Pull request by DerrickUnleashed**
This PR adds :
- Real-Time Detection Transformers Model (Object Detection Model)
- Test Suite for the same

Closes #327

### CODE REVIEW

**Comment by DerrickUnleashed** [comment]
```r
=======================================
Running inference with: model_rfdetr_base
========================================
Model weights for <rfdetr_base> (~123 MB) will be downloaded and processed if
not already available.
Loaded pretrained weights for <rfdetr_base> (487/487 keys, 0 skipped).
Top detections:
  score=0.8337 label=17 box=[156,37,1440,1148]
  score=0.1323 label=18 box=[156,37,1440,1148]
  score=0.0865 label=23 box=[156,37,1440,1148]
  score=0.0846 label=20 box=[156,37,1440,1148]
  score=0.0698 label=18 box=[164,23,1440,1148]
Detections above 0.3: 1

========================================
Running inference with: model_rfdetr_base_2
========================================
Model weights for <rfdetr_base_2> (~123 MB) will be downloaded and processed
if not already available.
Loaded pretrained weights for <rfdetr_base_2> (487/487 keys, 0 skipped).
Top detections:
  score=0.6605 label=17 box=[156,38,1440,1152]
  score=0.2869 label=18 box=[156,38,1440,1152]
  score=0.1364 label=23 box=[156,38,1440,1152]
  score=0.0740 label=20 box=[156,38,1440,1152]
  score=0.0676 label=16 box=[156,38,1440,1152]
Detections above 0.3: 1

========================================
Running inference with: model_rfdetr_base_o365
========================================
Model weights for <rfdetr_base_o365> (~127 MB) will be downloaded and
processed if not already available.
Loaded pretrained weights for <rfdetr_base_o365> (487/487 keys, 0 skipped).
Top detections:
  score=0.5500 label=140 box=[158,34,1440,1144]
  score=0.4891 label=93 box=[158,34,1440,1144]
  score=0.1077 label=343 box=[158,34,1440,1144]
  score=0.0971 label=321 box=[158,34,1440,1144]
  score=0.0838 label=84 box=[158,34,1440,1144]
Detections above 0.3: 2

========================================
Running inference with: model_rfdetr_large
========================================
Model weights for <rfdetr_large> (~518 MB) will be downloaded and processed
if not already available.
Loaded pretrained weights for <rfdetr_large> (533/533 keys, 0 skipped).
Top detections:
  score=0.9274 label=18 box=[156,40,1444,1136]
  score=0.0538 label=18 box=[163,36,1440,1137]
  score=0.0460 label=19 box=[156,40,1444,1136]
  score=0.0403 label=17 box=[163,36,1440,1137]
  score=0.0336 label=11 box=[156,40,1444,1136]
Detections above 0.3: 1

========================================
Running inference with: model_rfdetr_medium
========================================
Model weights for <rfdetr_medium> (~116 MB) will be downloaded and processed
if not already available.
Loaded pretrained weights for <rfdetr_medium> (465/465 keys, 0 skipped).
Top detections:
  score=0.3381 label=17 box=[155,35,1460,1158]
  score=0.2565 label=18 box=[155,35,1460,1158]
  score=0.1775 label=23 box=[155,35,1460,1158]
  score=0.1350 label=23 box=[157,34,1461,1156]
  score=0.0925 label=17 box=[157,34,1461,1156]
Detections above 0.3: 1

========================================
Running inference with: model_rfdetr_nano
========================================
Model weights for <rfdetr_nano> (~116 MB) will be downloaded and processed if
not already available.
Loaded pretrained weights for <rfdetr_nano> (465/465 keys, 0 skipped).
Top detections:
  score=0.3381 label=17 box=[155,35,1460,1158]
  score=0.2565 label=18 box=[155,35,1460,1158]
  score=0.1775 label=23 box=[155,35,1460,1158]
  score=0.1350 label=23 box=[157,34,1461,1156]
  score=0.0925 label=17 box=[157,34,1461,1156]
Detections above 0.3: 1

========================================
Running inference with: model_rfdetr_small
========================================
Model weights for <rfdetr_small> (~116 MB) will be downloaded and processed
if not already available.
Loaded pretrained weights for <rfdetr_small> (465/465 keys, 0 skipped).
Top detections:
  score=0.5835 label=17 box=[154,34,1459,1155]
  score=0.4582 label=18 box=[154,34,1459,1155]
  score=0.3348 label=23 box=[154,34,1459,1155]
  score=0.0870 label=64 box=[154,34,1459,1155]
  score=0.0798 label=20 box=[154,34,1459,1155]
Detections above 0.3: 3
```

fileef4133e485e8

File: NEWS.md
Comment (cregouby) [inline NEWS.md:1]: **improvement** Could you move the added line into the ## New models section ?

File: R/models-rfdetr_detection.R
Comment (cregouby) [inline R/models-rfdetr_detection.R:1]: **todo** **missing** Please add a representative example to the model documentation.
**todo** **missing** Please mention the attribution in a code comment `# This code is modified from ...`
**todo** Please fix merge conflicts

**Comment by DerrickUnleashed** [comment]
Can we implement a vignette for this perhaps? @cregouby

File: tests/testthat/test-models-rfdetr.R
Comment (cregouby) [inline tests/testthat/test-models-rfdetr.R:1]: **improvement** Could we be more specific on each tests through getting a little deeper than the output shape ? 
**suggestion** You should use the `expect_coco_model_detects_cat(model)` for each and every pretrained model. See code in https://github.com/mlverse/torchvision/blob/536a80a08618695b193249fe1f17f39dbbd982c4/tests/testthat/helper-torchvision.R#L69-L88
**question** Could we also test for `expect_bbox_to_be_xyxy()` from https://github.com/mlverse/torchvision/blob/536a80a08618695b193249fe1f17f39dbbd982c4/tests/testthat/helper-torchvision.R#L36-L65

File: R/models-rfdetr_detection.R [hunk 1]
```diff
@@ -0,0 +1,1617 @@
...
+)
+
+ms_deform_attn_core_pytorch <- function(value, spatial_shapes, sampling_locations, attention_weights,
+                                         spatial_shapes_hw = NULL) {
+  batch_size <- value$size(1)
+  n_heads <- value$size(2)
+  head_dim <- value$size(3)
+  len_query <- sampling_locations$size(2)
+  n_levels <- sampling_locations$size(4)
+  n_points <- sampling_locations$size(5)
```
Comment (cregouby) [inline R/models-rfdetr_detection.R]: **suggestion** you could make use of zeallot %<-% for readability

File: R/models-rfdetr_detection.R [hunk 2]
```diff
@@ -0,0 +1,1617 @@
...
+      pos <- list()
+      for (feat in feats) {
+        pos[[length(pos) + 1]] <- self[["1"]](feat, align_dim_orders = FALSE)
+      }
+      list(feats, pos)
+    }
+  }
+)
+
+ms_deform_attn_core_pytorch <- function(value, spatial_shapes, sampling_locations, attention_weights,
```
Comment (cregouby) [inline R/models-rfdetr_detection.R]: **thought** very strange name knowing that there is zero pytorch in the game... OK, this comes forme the code in roboflow repo. Then please mention the attribution as code coment.

File: R/models-rfdetr_detection.R [hunk 3]
```diff
@@ -0,0 +1,1617 @@
...
+    value <- value$transpose(2, 3)$contiguous()$view(c(
+      batch_size, self$n_heads, self$d_model %/% self$n_heads, -1
+    ))
+    output <- ms_deform_attn_core_pytorch(
+      value, input_spatial_shapes, sampling_locations, attention_weights,
+      input_spatial_shapes_hw
+    )
+    self$output_proj(output)
+  }
+)
```
Comment (cregouby) [inline R/models-rfdetr_detection.R]: **todo** please factorize this code with https://github.com/mlverse/torchvision/blob/8d19bace437fdb761b9e2b62668c9d0b96b37bc3/R/models-lw_detr.R#L315-L404
as the initialization is the same and  forward here only additionnaly  manages 2D  reference_points.
**suggestion** rename it `detr_ms_deform_attn`

File: R/models-rfdetr_detection.R [hunk 4]
```diff
@@ -0,0 +1,1617 @@
...
+      }
+      cls_enc <- torch_cat(cls_enc_list, dim = 2)
+      if (!is.null(hs)) {
+        out$enc_outputs <- list(pred_logits = cls_enc, pred_boxes = ref_enc)
+      } else {
+        out$pred_logits <- cls_enc
+        out$pred_boxes <- ref_enc
+      }
+    }
+    out
```
Comment (cregouby) [inline R/models-rfdetr_detection.R:1235]: **todo** Please make the output data model identical to all other objects detection models so that it helps visualization: we expect out to have a $detections with each detection item having names `c("boxes", "labels", "scores")`

File: R/models-rfdetr_detection.R [hunk 5]
```diff
@@ -0,0 +1,1617 @@
...
+  },
+  forward = function(x) {
+    for (i in seq_len(self$num_layers)) {
+      x <- self$layers[[i]](x)
+      if (i < self$num_layers) x <- nnf_relu(x)
+    }
+    x
+  }
+)
+
```
Comment (cregouby) [inline R/models-rfdetr_detection.R]: **suggestion** looks like this should be factorized  with / reused from  https://github.com/mlverse/torchvision/blob/8d19bace437fdb761b9e2b62668c9d0b96b37bc3/R/models-lw_detr.R#L409-L429

**Review by cregouby — CHANGES_REQUESTED** [review]
**praise** This is massive, thanks for it
**todo** see inline.

File: NEWS.md
Comment (DerrickUnleashed) [reply NEWS.md:1]: Yes I have added it in

File: R/models-rfdetr_detection.R
Comment (DerrickUnleashed) [reply R/models-rfdetr_detection.R:1]: Sure i have added a better example

File: tests/testthat/test-models-rfdetr.R
Comment (DerrickUnleashed) [reply tests/testthat/test-models-rfdetr.R:1]: I'll add a function so i'll be able to call it for each model making the readability better

File: R/models-rfdetr_detection.R (hunk 1 above)
Comment (DerrickUnleashed) [reply R/models-rfdetr_detection.R]: Sure I'll use zealott

File: R/models-rfdetr_detection.R (hunk 2 above)
Comment (DerrickUnleashed) [reply R/models-rfdetr_detection.R]: This may seem confusing, so I'll rename to ms_deform_attn_core.

File: R/models-rfdetr_detection.R (hunk 3 above)
Comment (DerrickUnleashed) [reply R/models-rfdetr_detection.R]: Sure I'll factorise it

File: R/models-rfdetr_detection.R (hunk 4 above)
Comment (DerrickUnleashed) [reply R/models-rfdetr_detection.R:1235]: Sure I have standardised the output format

File: R/models-rfdetr_detection.R (hunk 5 above)
Comment (DerrickUnleashed) [reply R/models-rfdetr_detection.R]: Sure will do also updated it to `detr_mlp_layers`

**Comment by DerrickUnleashed** [comment]
```r
url <- "https://upload.wikimedia.org/wikipedia/commons/6/6f/Toy_Poodle_wearing_clothes_in_Tokyo.jpg"
img <- base_loader(url) %>%
  transform_to_tensor()

model <- model_rfdetr_large(pretrained = TRUE)
model$eval()

pred <- torch::with_no_grad(
  model(img$unsqueeze(1))
)$detections[[1]]

topk <- pred$scores$topk(k = 5L)[[2]]
boxes <- pred$boxes[topk, ]
labels <- coco_classes(as.integer(pred$labels[topk]))
canvas <- (img * 255)$to(dtype = torch_uint8())
boxed <- draw_bounding_boxes(canvas, boxes, labels = labels,
 colors = "black", width = 6)
tensor_image_browse(boxed)
```

test
